README.md
megatools - command line client application for Mega
====================================================

Megatools is a collection of programs for accessing Mega service from 
a command line of your desktop or server.

Megatools allow you to copy individual files as well as entire directory 
trees to and from the cloud. You can also perform streaming downloads for 
example to preview videos and audio files, without needing to download 
the entire file.

Megatools are robust and optimized for fast operation - as fast as Mega 
servers allow. Memory requirements and CPU utilization are kept at minimum.

You can register account using a 'megareg' tool, with the benefit of having 
true control of your encryption keys.

Mega website can be found at http://mega.co.nz.

Megatools can be downloaded at http://megatools.megous.com


Tools
=====

  megareg      Register and verify a new mega account
  megadf       Show your cloud storage space usage/quota
  megals       List all remote files
  megamkdir    Create remote directory
  megarm       Remove remote file or directory
  megamv       Move and rename remote files
  megaput      Upload individual files
  megaget      Download individual files
  megadl       Download file from a "public" Mega link
               (doesn't require login)
  megastream   Streaming download of a file
               (can be used to preview videos or music)
  megasync     Upload or download a directory tree
  megafs       Mount remote filesystem locally.


All of these tools do:

- Local caching of remote session/filesystem information 
  for faster execution. Cache is encrypted with your password 
  key.
- Support loading login credentials from a configuration file


Usage
=====

See man pages for how to use individual tools and how to configure them:

  man megatools


Windows Quirks
==============

On Windows, if you see weird characters in your megals output, you'll need
to set correct CHARSET environment variable. For example on Czech Windows 
this would mean executing this command in cmd before using the tools:

  set CHARSET=CP852

This is just a cosmetic issue. Internally, megatools always work with UTF-8
file names, and even if the tool's terminal output is corrupted, files names
of downloaded/uploaded files will be correct.


Installation on Mac OS X (Mountain Lion)
========================================

Thanks to Carl Moden, megatools is available in Homebrew (http://brew.sh/).

You can therefrore install megatools with:

  brew install megatools

To install megatools by hand you need to:

1. Install glib-networking with a ca-bundle:

  export BUNDLE=/usr/local/share/ca-bundle.crt
  curl http://curl.haxx.se/ca/cacert.pem -o $BUNDLE
  /usr/bin/sed -ie "s|--without-ca-certificates|--with-ca-certificates=$BUNDLE|" /usr/local/Library/Formula/glib-networking.rb
  brew install glib-networking

2. Make the pkgconfig (.pc) file for libcurl findable by megatools configure script:

  export PKG_CONFIG_PATH=/usr/local/Library/ENV/pkgconfig/10.8

3. Untar megatools, run `./configure`, `make` and `make install` as usual.


Installation on Gentoo
======================

Third-party ebuild is available at:

  https://github.com/megabaks/stuff/tree/master/app-misc/megatools


Installation on Debian
======================

Megatools is available in sid thanks to Alberto Garcia:

  https://packages.debian.org/unstable/megatools


Installation on FreeBSD
=======================

Megatools is available in freshports thanks to Maxim V. Kostikov:

  http://www.freshports.org/net/megatools/


Building megatools from source code
===================================

You will need to install dependnencies first, to be able to do the build from
source code tarball.

On Debian, Ubuntu:

  apt-get -y install build-essential libglib2.0-dev libssl-dev libcurl4-openssl-dev libgirepository1.0-dev

On Fedora:

  yum -y install gcc make glib2-devel libcurl-devel openssl-devel gmp-devel tar

On OpenSUSE:

  zypper -n install gcc make glib2-devel libcurl-devel openssl-devel gmp-devel

On Arch Linux:

  pacman -Sy --noconfirm --needed pkg-config gcc make glib2 curl gmp nettle


Author
======

Megatools were written by Ondřej Jirman <megous@megous.com>, 2014
Webiste: http://megatools.megous.com


Support and bug reports
=======================

If you think you've found bug in megatools, send a report including enough
information for recreating the issue to: megous@megous.com


License
=======

Megatools are licensed under GPLv2 with OpenSSL exemption, see LICENSE
file for details.

This product includes software developed by the OpenSSL Project for use
in the OpenSSL Toolkit. (http://www.openssl.org/)


NEWS
megatools 1.9.94 - 2015-01-02
=============================

Bugfix and optimization release.

Fixes:
- Force use of IPv4 addresses
- Fix path sanitization so that trailing slashes are ignored
- Document config file oddities in megarc manpage
- Add info on how to build/install megatools on various distros

Improvements:
- Optimize path handling code so that filesystems with many files are 
  usable. Previously, megatools would hang for quite some time when
  updating it's path map.
- Use non-recursive makefile to build megatools

megatools 1.9.93 - 2014-10-24
=============================

Added OpenSSL exemption to the license.

Fixes:

- megadf now returns correct values
- fixed path sanitization in mega_session_stat
- raised safety limit os size of POST responses from 32 to 256 MiB
- show progress when downloading folder using megadl
- optimized buffer allocation during donwload and upload
- report details of errors when stream callback fails to read/write file
- fix detection of convert executeable for preview generation
- allow to disable file previews

megatools 1.9.92 - 2014-08-24
=============================

Created stable branch for bugfixes. This is the release from the stable 
branch.

This release contains mostly small bug fixes. Most important change is
that Megatools now recognizes server busy errors as reported by mega.co.nz
via HTTP status 500 Server Is Busy, and retries the communication. This
reduces command failures by a lot, and you'll not see 500 Server Is Busy 
errors anymore.

Another major news is update to the windows package. There are now two 
packages for 32bit and 64bit windows platform, and all dlls were updated to
their latest versions.

Other news:

- Added 64bit windows package
- Unmask Megatools to Mega.co.nz (user Megatools/VERSION User-Agent)
- Check for a2x if --enable-docs-build=yes
- Megadl now reports file download errors
- Add megadl --print-names option
- Fix megasync error (issue #45 on github)
- Fix megadl used with pipe (--path -) on windows (issue #38)
- Installation instructions for Mac
- Add info about Gentoo ebuild to the README file (issue #26)
- Fix Megatools thumbnails for PDF (issue #5)
- Do repeated API calls after 500 Server Is Busy and increase retry delay
- Fix issue to download a file with accent with megaget (issue #22)

megatools 1.9.91 - 2013-04-14
=============================

Please note, that megatools is using GNOME release versioning. That means
that in MAJOR.MINOR.PATCH, odd MINOR number signifies development version
of the tools. PATCH version is increased on each public release.

By using odd MINOR versions of the software you are using an unstable
branch. That doesn't mean it will be crashing on you, but it may have 
certain quirks, and oddities, missing or incomplete features, etc.

Next stable release is going to be 2.0.0.

To the news:

- Added man pages for all tools
- Removed megastream in favor of mega{dl,get} --path -
- Megadf - fix unsigned overflow when server reports 0 total space
- Implemented --debug option
- Win32: Drop requirement for manual setup of GIO_EXTRA_MODULES
- Load certificates from the binaries folder on windows
- Enable CA validation for TLS connections
- Fix compilation on FreeBSD by adding --enable-warnings
- Add note that on FreeBSD GNU make must be used to build the tools
- Fix typedef redefinitions (error on older gcc's)
- Fix megals --names logic
- megaput/megaget/megadl/megasync: --no-progress support
- Fix mega_session_dl streaming mode segfault
- megamkdir/megarm: Check for required arguments
- Pretty print JSON in debug output

AUR package is available at https://aur.archlinux.org/packages/megatools/

Ubunut PPA is availble at https://launchpad.net/~megous/+archive/ppa

If someone wants to take over packaging for Ubuntu or Debian, that would
be very appretiated, as my speciality is Arch Linux.


megatools 1.9.90 - 2013-04-11
=============================

Implemented much simpler and faster json parser than json-glib, which was
used before. Megatools now use sjson, which you can find at:

  https://github.com/megous/sjson

Json-glib dependency was thus dropped.

Initial work on getting rid of dependency on libcurl.

Initial work on the new libmega library with support for gobject-introspection
based bindings to various languages: JavaScript, Python, Ruby, LUA,...

Experimental implementation of Mega API in JavaScript (gjs), this should help
people develop GNOME GUI for Mega in the future.

New command line tools:

  megafs  - mount mega.co.nz filesystem usong FUSE (Linux only)
  megamv  - move and rename remote files and directories


megatools 1.2.2 - 2013-04-11
============================

Bugs fixed:

- Trash and Inbox paths are no longer mixed up (bug reported by Ovidiu)
- megadl segfault
- megaput --path /Root/nonexistent-file local-file (remote file would get 
  the name of the local file, instead of nonexistent-file)
- megals /Root/file-name would return nothing
- Segfault with password length over 15 chars
- Fix file size parsing to make it work with files over 2GB in size


megatools 1.2.1 - 2013-03-12
============================

Added support for OpenSUSE 12.1 and newer versions.

Build libmega statically for binary distribution to ease installation 
outside of /usr/local.


megatools 1.2.0 - 2013-03-12
============================

Megatools now support download of shared files and folders.

Megatools support creating thumbnails for image and video files uploaded 
to Mega. This is enabled only on Linux, as it requires ffmpegthumbnailer 
and convert (from ImageMagick package).

You can now manage your contacts list via:
  
  - mkdir /Contacts/some@email.com
  - rm /Contacts/some.other@email.com

Exported folders can be downloaded using megadl.

All tools gained a few more options to help with scripting needs.

If you don't provide password on command-line or in config file you'll 
be asked for it when the tool runs.

Megals tool gained many new options.


libtools/tools.c
/*
 *  megatools - Mega.co.nz client library and tools
 *  Copyright (C) 2013  Ondřej Jirman <megous@megous.com>
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License along
 *  with this program; if not, write to the Free Software Foundation, Inc.,
 *  51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
 */

#include <curl/curl.h>
#include <openssl/err.h>
#include <openssl/ssl.h>
#include <locale.h>

#include "config.h"
#include "tools.h"

#ifdef G_OS_WIN32
#include <windows.h>
#else
#include <signal.h>
#include <termios.h>
#include <unistd.h>
#endif

#ifdef G_OS_WIN32
#define MEGA_RC_FILENAME "mega.ini"
#else
#define MEGA_RC_FILENAME ".megarc"
#endif

static GOptionContext* opt_context;
static gchar* opt_username;
static gchar* opt_password;
static gchar* opt_config;
static gboolean opt_reload_files;
static gint opt_cache_timout = 10 * 60;
static gboolean opt_version;
static gboolean opt_no_config;
static gboolean opt_no_ask_password;
static gboolean opt_disable_previews;
gboolean tool_allow_unknown_options = FALSE;

static gboolean opt_debug_callback(const gchar *option_name, const gchar *value, gpointer data, GError **error)
{
  if (value)
  {
    gchar** opts = g_strsplit(value, ",", 0);
    gchar** opt = opts;

    while (*opt)
    {
      if (g_ascii_strcasecmp(*opt, "api") == 0)
        mega_debug |= MEGA_DEBUG_API;
      else if (g_ascii_strcasecmp(*opt, "fs") == 0)
        mega_debug |= MEGA_DEBUG_FS;
      else if (g_ascii_strcasecmp(*opt, "cache") == 0)
        mega_debug |= MEGA_DEBUG_CACHE;

      opt++;
    }
  }
  else
  {
    mega_debug = MEGA_DEBUG_API;
  }

  return TRUE;
}

static GOptionEntry basic_options[] =
{
  { "debug",              '\0',  G_OPTION_FLAG_OPTIONAL_ARG, G_OPTION_ARG_CALLBACK, opt_debug_callback, "Enable debugging output",  "OPTS"  },
  { "version",            '\0',  0,                          G_OPTION_ARG_NONE,     &opt_version,       "Show version information", NULL    },
  { NULL }
};

static GOptionEntry auth_options[] =
{
  { "username",            'u',  0, G_OPTION_ARG_STRING,    &opt_username,         "Account username (email)",                    "USERNAME" },
  { "password",            'p',  0, G_OPTION_ARG_STRING,    &opt_password,         "Account password",                            "PASSWORD" },
  { "config",             '\0',  0, G_OPTION_ARG_FILENAME,  &opt_config,           "Load configuration from a file",              "PATH"     },
  { "ignore-config-file", '\0',  0, G_OPTION_ARG_NONE,      &opt_no_config,        "Disable loading " MEGA_RC_FILENAME,           NULL       },
  { "no-ask-password",    '\0',  0, G_OPTION_ARG_NONE,      &opt_no_ask_password,  "Never ask interactively for a password",      NULL       },
  { "disable-previews",   '\0',  0, G_OPTION_ARG_NONE,      &opt_disable_previews, "Never generate previews when uploading file", NULL       },
  { "reload",             '\0',  0, G_OPTION_ARG_NONE,      &opt_reload_files,     "Reload filesystem cache",                     NULL       },
  { NULL }
};

#if GLIB_CHECK_VERSION(2, 32, 0)

static GMutex* openssl_mutexes = NULL;

static void openssl_locking_callback(int mode, int type, const char *file, int line)
{
  if (mode & CRYPTO_LOCK)
    g_mutex_lock(openssl_mutexes + type);
  else
    g_mutex_unlock(openssl_mutexes + type);
}

static unsigned long openssl_thread_id_callback()
{
  unsigned long ret;
  ret = (unsigned long)g_thread_self();
  return ret;
}

static void init_openssl_locking()
{
  gint i;

  // initialize OpenSSL locking for multi-threaded operation
  openssl_mutexes = g_new(GMutex, CRYPTO_num_locks());
  for (i = 0; i < CRYPTO_num_locks(); i++)
    g_mutex_init(openssl_mutexes + i);

  SSL_library_init();
  CRYPTO_set_id_callback(openssl_thread_id_callback);
  CRYPTO_set_locking_callback(openssl_locking_callback);
}

#else

static GMutex** openssl_mutexes = NULL;

static void openssl_locking_callback(int mode, int type, const char *file, int line)
{
  if (mode & CRYPTO_LOCK)
    g_mutex_lock(openssl_mutexes[type]);
  else
    g_mutex_unlock(openssl_mutexes[type]);
}

static unsigned long openssl_thread_id_callback()
{
  unsigned long ret;
  ret = (unsigned long)g_thread_self();
  return ret;
}

static void init_openssl_locking()
{
  gint i;

  // initialize OpenSSL locking for multi-threaded operation
  openssl_mutexes = g_new(GMutex*, CRYPTO_num_locks());
  for (i = 0; i < CRYPTO_num_locks(); i++)
    openssl_mutexes[i] = g_mutex_new();

  SSL_library_init();
  CRYPTO_set_id_callback(openssl_thread_id_callback);
  CRYPTO_set_locking_callback(openssl_locking_callback);
}

#endif

#ifdef G_OS_WIN32
static gchar* get_tools_dir(void)
{
  gchar *path = NULL;
  wchar_t *wpath;
  DWORD len = PATH_MAX;

  HMODULE handle = GetModuleHandleW(NULL);

  wpath = g_new0(wchar_t, len);
  if (GetModuleFileNameW(handle, wpath, len) < len)
    path = g_utf16_to_utf8(wpath, -1, NULL, NULL, NULL);
  g_free(wpath);

  if (path == NULL)
    path = g_strdup("");

  gchar* dir = g_path_get_dirname(path);
  g_free(path);
  return dir;
}
#endif

static void init(void)
{
#if !GLIB_CHECK_VERSION(2, 32, 0)
  if (!g_thread_supported())
    g_thread_init(NULL);
#endif

  setlocale(LC_ALL, "");

#if !GLIB_CHECK_VERSION(2, 36, 0)
  g_type_init();
#endif

#ifndef G_OS_WIN32
  signal(SIGPIPE, SIG_IGN);
#endif

  init_openssl_locking();

#ifdef G_OS_WIN32
  gchar* tools_dir = get_tools_dir();

  gchar* tmp = g_build_filename(tools_dir, "gio", NULL);
  g_setenv("GIO_EXTRA_MODULES", tmp, TRUE);
  g_free(tmp);

  gchar* certs = g_build_filename(tools_dir, "ca-certificates.crt", NULL);
  g_setenv("CA_CERT_PATH", certs, TRUE);
  g_free(certs);

  g_free(tools_dir);
#endif
}

static gchar* input_password(void)
{
  gint tries = 3;
  gchar buf[256];
  gchar* password = NULL;

#ifdef G_OS_WIN32
  HANDLE hStdin = GetStdHandle(STD_INPUT_HANDLE); 
  DWORD mode = 0;
  GetConsoleMode(hStdin, &mode);
  SetConsoleMode(hStdin, mode & (~ENABLE_ECHO_INPUT));
#else
  struct termios oldt;
  tcgetattr(STDIN_FILENO, &oldt);
  struct termios newt = oldt;
  newt.c_lflag &= ~ECHO;
  tcsetattr(STDIN_FILENO, TCSANOW, &newt);
#endif

again:
  g_print("Enter password for (%s): ", opt_username);
  if (fgets(buf, 256, stdin))
  {
    if (strlen(buf) > 1)
    {
      password = g_strndup(buf, strcspn(buf, "\r\n"));
    }
    else
    {
      if (--tries > 0)
      {
        g_print("\n");
        goto again;
      }

      g_print("\nYou need to provide non-empty password!\n");
      exit(1);
    }
  }
  else
  {
    g_printerr("\nERROR: Can't read password from the input!\n");
    exit(1);
  }

#ifdef G_OS_WIN32
  SetConsoleMode(hStdin, mode);
#else
  tcsetattr(STDIN_FILENO, TCSANOW, &oldt);
#endif

  g_print("\nGood, signing in...\n");

  return password;
}

static void print_version(void)
{
  if (opt_version)
  {
    g_print("megatools " VERSION " - command line tools for Mega.co.nz\n\n");
    g_print("Written by Ondřej Jirman <megous@megous.com>, 2013\n");
    g_print("Go to http://megatools.megous.com for more information\n");
    exit(0);
  }
}

gchar* tool_convert_filename(const gchar* path, gboolean local)
{
  gchar* locale_path;

#ifdef G_OS_WIN32
  locale_path = g_locale_to_utf8(path, -1, NULL, NULL, NULL);
#else
  if (local)
    locale_path = g_strdup(path);
  else
    locale_path = g_locale_to_utf8(path, -1, NULL, NULL, NULL);
#endif

  if (locale_path == NULL)
  {
    g_printerr("ERROR: Invalid filename locale, can't convert file names specified on the command line to UTF-8.\n");
    exit(1);
  }

  return locale_path;
}

void tool_init_bare(gint* ac, gchar*** av, const gchar* tool_name, GOptionEntry* tool_entries)
{
  GError *local_err = NULL;

  init();

  opt_context = g_option_context_new(tool_name);
  if (tool_allow_unknown_options)
    g_option_context_set_ignore_unknown_options(opt_context, TRUE);
  if (tool_entries)
    g_option_context_add_main_entries(opt_context, tool_entries, NULL);
  g_option_context_add_main_entries(opt_context, basic_options, NULL);

  if (!g_option_context_parse(opt_context, ac, av, &local_err))
  {
    g_printerr("ERROR: Option parsing failed: %s\n", local_err->message);
    g_clear_error(&local_err);
    exit(1);
  }

  print_version();
}

void tool_init(gint* ac, gchar*** av, const gchar* tool_name, GOptionEntry* tool_entries)
{
  GError *local_err = NULL;

  init();

  opt_context = g_option_context_new(tool_name);
  if (tool_allow_unknown_options)
    g_option_context_set_ignore_unknown_options(opt_context, TRUE);
  if (tool_entries)
    g_option_context_add_main_entries(opt_context, tool_entries, NULL);
  g_option_context_add_main_entries(opt_context, auth_options, NULL);
  g_option_context_add_main_entries(opt_context, basic_options, NULL);

  if (!g_option_context_parse(opt_context, ac, av, &local_err))
  {
    g_printerr("ERROR: Option parsing failed: %s\n", local_err->message);
    g_clear_error(&local_err);
    exit(1);
  }

  print_version();

  // load username/password from ini file
  if (!opt_no_config || opt_config)
  {
    gboolean status;
    gc_key_file_unref GKeyFile* kf = g_key_file_new();

    if (opt_config)
      status = g_key_file_load_from_file(kf, opt_config, 0, NULL);
    else
    {
      status = g_key_file_load_from_file(kf, MEGA_RC_FILENAME, 0, NULL);
      if (!status)
      {
        gc_free gchar* tmp = g_build_filename(g_get_home_dir(), MEGA_RC_FILENAME, NULL);
        status = g_key_file_load_from_file(kf, tmp, 0, NULL);
      }
    }

    if (status)
    {
      if (!opt_username)
        opt_username = g_key_file_get_string(kf, "Login", "Username", NULL);
      if(!opt_password)
        opt_password = g_key_file_get_string(kf, "Login", "Password", NULL);

      gint to = g_key_file_get_integer(kf, "Cache", "Timeout", &local_err);
      if (local_err == NULL)
        opt_cache_timout = to;
      else
        g_clear_error(&local_err);
    }
  }

  if (!opt_username)
  {
    g_printerr("ERROR: You must specify your mega.co.nz username (email)\n");
    exit(1);
  }

  if (!opt_password && opt_no_ask_password)
  {
    g_printerr("ERROR: You must specify your mega.co.nz password\n");
    exit(1);
  }

  if (!opt_password)
    opt_password = input_password();
}

mega_session* tool_start_session(void)
{
  GError *local_err = NULL;
  gchar* sid = NULL;
  gboolean loaded = FALSE;

  mega_session* s = mega_session_new();

  // try to load cached session data (they are valid for 10 minutes since last
  // user_get or refresh)
  if (!mega_session_load(s, opt_username, opt_password, opt_cache_timout, &sid, &local_err))
  {
    g_clear_error(&local_err);

    if (!mega_session_open(s, opt_username, opt_password, sid, &local_err))
    {
      g_printerr("ERROR: Can't login to mega.co.nz: %s\n", local_err->message);
      goto err;
    }

    if (!mega_session_refresh(s, &local_err))
    {
      g_printerr("ERROR: Can't read filesystem info from mega.co.nz: %s\n", local_err->message);
      goto err;
    }

    loaded = TRUE;
    mega_session_save(s, NULL);
  }

  if (opt_reload_files && !loaded)
  {
    if (!mega_session_refresh(s, &local_err))
    {
      g_printerr("ERROR: Can't read filesystem info from mega.co.nz: %s\n", local_err->message);
      goto err;
    }

    mega_session_save(s, NULL);
  }

  mega_session_enable_previews(s, !opt_disable_previews);

  g_free(sid);
  return s;

err:
  mega_session_free(s);
  g_clear_error(&local_err);
  g_free(sid);
  return NULL;
}

void tool_fini(mega_session* s)
{
  if (s)
    mega_session_free(s);

  g_option_context_free(opt_context);
  curl_global_cleanup();
  CRYPTO_cleanup_all_ex_data();
  ERR_free_strings();

  CRYPTO_set_id_callback(NULL);
  CRYPTO_set_locking_callback(NULL);
  g_free(openssl_mutexes);
}
